src/python/pants/backend/swift/goals/check.py

```diff
@@ -0,0 +1,74 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+from itertools import groupby
+from typing import Iterable
+
+from pants.backend.swift.subsystems.toolchain import SwiftSubsystem
+from pants.backend.swift.target_types import SwiftFieldSet, SwiftSourceField
+from pants.backend.swift.util_rules.compile import (
+    FallibleTypecheckedSwiftModule,
+    TypecheckSwiftModuleRequest,
+)
+from pants.core.goals.check import CheckRequest, CheckResult, CheckResults
+from pants.engine.rules import Get, MultiGet, Rule, collect_rules, rule
+from pants.engine.target import AllTargets, AllTargetsRequest
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+
+logger = logging.getLogger(__name__)
+
+
+class SwiftCheckRequest(CheckRequest):
+    field_set_type = SwiftFieldSet
+    name = SwiftSubsystem.options_scope
+
+
+@rule(desc="Typecheck Swift compilation", level=LogLevel.DEBUG)
+async def swift_typecheck(request: SwiftCheckRequest) -> CheckResults:
+    # Retrieve the sources targets (swift modules) represented by the check request
+    # Checking a single-file could fail, because it's missing context from other files in the module - so typecheck the module
+    # More importantly, changing a single file could cause failures in other files in that module
+
+    # From the list of field sets, extract a single address per target (to later retrieve the target)
+    requested_target_names = {field_set.address.target_name for field_set in request.field_sets}
+
+    # TODO: There is no way this is correct - there must be a simpler way to grab all the swift_sources
```

Comment:
Given:
```swift
swift_sources(name="lib", sources=["1.swift", "2.swift", "3.swift"])
```

If I run `./pants check 2.swift` - I need to retrieve all three of those files in order to compile the module and avoid type-checking errors. I thought I was doing this correctly before, but then I realized that I was only picking some of the generated source fields, but not all of them.

Reply:
So, is it the case that all of the files in a single directory are always a single swift module? If so, then rather than breaking up into separate file targets and then merging again for typechecking, you might not want to use target generation at all, and only have `swift_sources` (no `swift_source` target).

On the other hand, if it _is_ possible to compile individual files as long as their explicit dependencies are present, then rather than merging files explicitly like this, you'll likely want to go ahead and implement dependency inference. That would cause any relevant files within a directory to be provided as dependencies here (and if they had cycles, you'd want to request them using `CoarsenedTarget` the way that the JVM backend does).

Reply:
I discussed this at length with Josh, and it made me realize more and more how much I dislike how Swift has handled modularization. I didn't like it back in the day, and it's kinda gotten more annoying to me.

For an example of how this can be used:
https://github.com/RobotPajamas/SwiftyTeeth/blob/master/Package.swift
https://github.com/RobotPajamas/SwiftyTeeth/tree/master/Sources

> So, is it the case that all of the files in a single directory are always a single swift module

When using Swift Package Manager (SPM) and Package.swift, you point at your module's source directory, and it assumes everything within that folder is part of the module. However, it does this by recursively grabbing each file individually and compiling it together with `swiftc`. In my case, inside of `SwiftyTeeth` and `SwiftyTooth`, I symlink down to my `Common` directory - just to fit within the SPM paradigm of 1 folder to 1 module. 

There is nothing that dictates that methodology is how it works for the Swift language in general, it's more like a convention for SPM - since I can otherwise put files anywhere on my filesystem, and manually/tediously compile them together using `swiftc`.

I'm 100% okay with doing the same (similar to `Go`) and just saying that you point `swift_sources` at a folder, and then we recursively grab everything within it. However, this mechanism doesn't follow the typical Pants 1 BUILD file per folder concept. 

>  On the other hand, if it is possible to compile individual files as long as their explicit dependencies are present, then rather than merging files explicitly like this, you'll likely want to go ahead and implement dependency inference.

I don't know much about Pant's dependency inference works, unfortunately. If this were C/C++, I would look at `include`(s) and then maybe handle discovery that way. 

However, with Swift, it's just a big bag of files that get compiled together, and all functions become available (depending on visibility attributes). No explicit imports of same-module files. 

Given that, I don't know how to cleanly infer from source code. I'm spending some time looking through the `swiftc` options to figure out if there is an option to emit the symbols it expects (and maybe can't find), and then look through the associated files to find it?

I think critically, this still requires a root directory to point at, so the use case where the files you want to compile are in a sub-folder - well, we don't really get a clean 1:1:1.

> That would cause any relevant files within a directory to be provided as dependencies here

As in. If I could get dep inference working, each source file would have a generated `dependencies` field consisting of the inferred files? Am I re-stating that correctly?


In terms of next steps, I have two options I think:

1. Go with the 80/20 solution of using a directory as the base item for Swift compilation (and recursively adding files?). It could even have a new target generator name (`swift_module`) which creates a bunch of `swift_source` files. 
2. Spend more time with dep inference, in the hopes that I can find a way to determine what is required for each compilation (however, I would still need to parse the `swift_sources` or supply a list of directories with potential source files).

Reply:
And, before I push any more code - I think the new plan will be to align with the official Swift terminology - which actually brings this backend closer to Go.

https://www.swift.org/package-manager/#conceptual-overview
> Swift organizes code into modules. Each module specifies a namespace and enforces access controls on which parts of that code can be used outside of the module.

> A package consists of Swift source files and a manifest file. The manifest file, called Package.swift, defines the package’s name and its contents using the PackageDescription module.

So, that would mean changing `tailor` rules to create a `swift_package` in the folders which have a `Package.swift` and then all of the source files are bundled into modules. 

Swift then uses "products" to mean libraries and executables - but it's probably perfectly fair to stick with `swift_library` and `swift_binary` or `swift_executable` as that's what products actually group together into one concept.